### PR TITLE
Subaru: Crosstrek Hybrid dashcam mode

### PR DESCRIPTION
**Pull request by adeebshihadeh**

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
@martinl got a test route for me?

File: tools/plotjuggler/layouts/torque-controller.xml [hunk 1]
```diff
@@ -55,6 +55,12 @@
   <plugin ID="Cereal Subscriber"/>
  </Plugins>
  <!-- - - - - - - - - - - - - - - -->
+ <previouslyLoaded_Datafiles>
+  <fileInfo filename="../../../../threepilot/tools/plotjuggler/tmp03i8v25j.rlog" prefix="">
+   <selected_datasources value=""/>
+   <plugin ID="DataLoad Rlog"/>
+  </fileInfo>
+ </previouslyLoaded_Datafiles>
```
Comment (sshane) [inline tools/plotjuggler/layouts/torque-controller.xml:63]:
```suggestion
```

**Comment by sshane** [comment]
It appears there are no similar messages on bus 0 to the cruise active status. Most common bits are quite a bit off:

```python
hex(msg)='0x361', bit_mismatches=981 of 3720 (26.37%), byt_idx=1, bit_idx=2
hex(msg)='0x1f9', bit_mismatches=1476 of 6009 (24.56%), byt_idx=5, bit_idx=7
hex(msg)='0x1f9', bit_mismatches=1002 of 6009 (16.67%), byt_idx=5, bit_idx=6
hex(msg)='0x244', bit_mismatches=1589 of 10190 (15.59%), byt_idx=4, bit_idx=3
Searched bus: 0
```

On bus 2, it gets a bit interesting. There's of course the bit on 0x321 which is what I'm comparing (ES_DashStatus), but also 0x220 which is (ES_Brake):

```python
ex(msg)='0x321', bit_mismatches=864 of 2423 (35.66%), byt_idx=5, bit_idx=2
hex(msg)='0x220', bit_mismatches=56 of 4800 (1.17%), byt_idx=0, bit_idx=0
hex(msg)='0x220', bit_mismatches=56 of 4800 (1.17%), byt_idx=4, bit_idx=0
hex(msg)='0x321', bit_mismatches=9 of 2423 (0.37%), byt_idx=4, bit_idx=3
Searched bus: 2
```

However I checked both signals on older ICE cars, and they proved to be inaccurate, specifically:

- `ES_DashStatus->Cruise_Activated` on Forester 2019 [goes low when at a standstill](https://user-images.githubusercontent.com/25857203/186993568-6474c94f-c12b-4162-9d1d-45011b71563d.png)
- `ES_DashStatus->Cruise_Activated` on Subaru Impreza 2019 goes low [when user is actuating the gas pedal while ACC stays engaged](https://user-images.githubusercontent.com/25857203/186994197-de10e5d8-0737-4307-807b-851d641a967e.png)
- `ES_Brake->Cruise_Activated` on Subaru Forester 2019 stays high [after user disengaged with the brake at a standstill](https://user-images.githubusercontent.com/25857203/186994577-57683709-4aa0-433d-8614-d818c402de73.png)

There's no other similar signals to `ES_DashStatus->Cruise_Activated`, so we'll have to verify that the signal on the hybrids don't have these problems on the ICE cars.

**Comment by sshane** [comment]
This is also missing the hybrid signals for gas and brake: https://github.com/ClockeNessMnstr/openpilot/blob/crosstrek-pid/selfdrive/car/subaru/carstate.py#L23

**Comment by martinl** [comment]
There is also ES_Status (0x222) Cruise_Activated, I'll check if it behaves differently

**Comment by martinl** [comment]
On Crosstrek ICE, ES_Status Cruise_Activated is identical to CruiseControl Cruise_Activated signal, I tested with 3 cases above - stopping behind lead car / hold (standstill), disengage using brake while in standstill, gas press while engaged
`8de015561e1ea4a0|2022-12-10--14-08-26`

**Comment by martinl** [comment]
I checked Crosstrek Hybrid dbc and it does not have ES_Status message. I'll find the signal once I find someone who can provide route for Crosstrek Hybrid

**Comment by jnewb1** [comment]
crosstrek hybrid route: 31d10447d8c7dc19|2023-08-17--07-30-31--0

**Comment by jnewb1** [comment]
I'll just merge the hybrid cars behind dashcam, we still need a good cruise_activated bit. Would be helpful to get a forester hybrid route so I can see if they are the same

**Comment by sshane** [comment]
Can you merge https://github.com/commaai/openpilot/pull/24770 after this? Make sure to leave a comment at least of what is required to un-dashcam

File: selfdrive/car/subaru/carcontroller.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 from openpilot.selfdrive.car import apply_driver_steer_torque_limits
 from openpilot.selfdrive.car.subaru import subarucan
 from openpilot.selfdrive.car.subaru.values import DBC, GLOBAL_GEN2, PREGLOBAL_CARS, CanBus, CarControllerParams, SubaruFlags
+from selfdrive.car.subaru.values import HYBRID_CARS
```
Comment (sshane) [inline selfdrive/car/subaru/carcontroller.py:6]: we already import from values above

File: selfdrive/car/subaru/carstate.py [hunk 3]
```diff
@@ -87,63 +95,77 @@ def update(self, cp, cp_cam, cp_body):
...
       cp_es_status = cp_body if self.car_fingerprint in GLOBAL_GEN2 else cp_cam
-      self.es_status_msg = copy.copy(cp_es_status.vl["ES_Status"])
-      self.cruise_control_msg = copy.copy(cp_cruise.vl["CruiseControl"])
-      self.brake_status_msg = copy.copy(cp_brakes.vl["Brake_Status"])
 
-    self.es_distance_msg = copy.copy(cp_es_distance.vl["ES_Distance"])
+      if self.car_fingerprint not in HYBRID_CARS: # Hybrid cars don't have es_distance, need a replacement
+        # 8 is known AEB, there are a few other values related to AEB we ignore
+        ret.stockAeb = (cp_es_distance.vl["ES_Brake"]["AEB_Status"] == 8) and \
+                      (cp_es_distance.vl["ES_Brake"]["Brake_Pressure"] != 0)
```
Comment (sshane) [inline selfdrive/car/subaru/carstate.py:107]: spacing

File: selfdrive/car/subaru/carstate.py [hunk 4]
```diff
@@ -87,63 +95,77 @@ def update(self, cp, cp_cam, cp_body):
...
       self.es_lkas_state_msg = copy.copy(cp_cam.vl["ES_LKAS_State"])
       cp_es_brake = cp_body if self.car_fingerprint in GLOBAL_GEN2 else cp_cam
       self.es_brake_msg = copy.copy(cp_es_brake.vl["ES_Brake"])
       cp_es_status = cp_body if self.car_fingerprint in GLOBAL_GEN2 else cp_cam
-      self.es_status_msg = copy.copy(cp_es_status.vl["ES_Status"])
-      self.cruise_control_msg = copy.copy(cp_cruise.vl["CruiseControl"])
-      self.brake_status_msg = copy.copy(cp_brakes.vl["Brake_Status"])
 
-    self.es_distance_msg = copy.copy(cp_es_distance.vl["ES_Distance"])
+      if self.car_fingerprint not in HYBRID_CARS: # Hybrid cars don't have es_distance, need a replacement
```
Comment (sshane) [inline selfdrive/car/subaru/carstate.py:104]: comment spacing

File: selfdrive/car/subaru/carstate.py [hunk 5]
```diff
@@ -87,63 +95,77 @@ def update(self, cp, cp_cam, cp_body):
...
+        # 8 is known AEB, there are a few other values related to AEB we ignore
+        ret.stockAeb = (cp_es_distance.vl["ES_Brake"]["AEB_Status"] == 8) and \
+                      (cp_es_distance.vl["ES_Brake"]["Brake_Pressure"] != 0)
+        self.es_distance_msg = copy.copy(cp_es_distance.vl["ES_Distance"])
+
+        self.es_status_msg = copy.copy(cp_es_status.vl["ES_Status"])
+        self.cruise_control_msg = copy.copy(cp_cruise.vl["CruiseControl"])
+
+    if self.car_fingerprint not in HYBRID_CARS:
+      self.es_distance_msg = copy.copy(cp_es_distance.vl["ES_Distance"])
```
Comment (sshane) [inline selfdrive/car/subaru/carstate.py:114]: seems like we can simplify this so we don't need to copy twice?

**Comment by sshane** [comment]
Make sure to request a final review before merging

**Pull request merged**